streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java

```diff
@@ -168,29 +171,59 @@ public void run() {
                 ProducerRecord<byte[], byte[]> record =
                         new ProducerRecord<>("data", stringSerde.serializer().serialize("", key), intSerde.serializer().serialize("", value));
 
-                producer.send(record, new Callback() {
-                    @Override
-                    public void onCompletion(final RecordMetadata metadata, final Exception exception) {
-                        if (exception != null) {
-                            exception.printStackTrace();
-                            Exit.exit(1);
-                        }
-                    }
-                });
-
+                producer.send(record, new TestCallback(record, needRetry));
 
                 numRecordsProduced++;
                 allData.get(key).add(value);
                 if (numRecordsProduced % 100 == 0)
                     System.out.println(numRecordsProduced + " records produced");
                 Utils.sleep(2);
+            }
+        }
+        producer.flush();
 
+        int remainingRetries = 5;
+        while (!needRetry.isEmpty()) {
+            final List<ProducerRecord<byte[], byte[]>> needRetry2 = new ArrayList<>();
+            for (final ProducerRecord<byte[], byte[]> record : needRetry) {
+                producer.send(record, new TestCallback(record, needRetry2));
+            }
+            producer.flush();
```

Comment:
Even though we are calling flush here is it possible we could end up back at the top of the loop before all `TestCallback. onCompletion`s have fired?  Seems to me we need to add some sort of check after re-assigning the `needRetry2` list, but I don't have any ideas ATM.

Reply:
From my understanding, `flush()` blocks until all buffered records are completely processed, including firing all corresponding callbacks. Compare the JavaDoc. \cc @hachikuji @apurvam (can you confirm)

Reply:
Your understanding is correct @mjsax .